README.md
# angular-qr v0.1.4 [![Build Status](https://travis-ci.org/janantala/angular-qr.png?branch=master)](https://travis-ci.org/janantala/angular-qr)

QR code generator for AngularJS

### Demo

Check out http://janantala.github.io/angular-qr/demo/

# Requirements

- AngularJS v 1.0+
- [qrcode.js](https://github.com/janantala/qrcode) (`bower install qrcode` is installed with angular-qr as dependecy)

# Usage

We use [bower](http://twitter.github.com/bower/) for dependency management. Add

```json
dependencies: {
    "angular-qr": "latest"
}
```

To your `bower.json` file. Then run

    bower install

This will copy the angular-qr files into your `bower_components` folder, along with its dependencies. Load the script files in your application:

```html
<script type="text/javascript" src="bower_components/angular/angular.js"></script>
<script type="text/javascript" src="bower_components/qrcode/lib/qrcode.min.js"></script>
<script type="text/javascript" src="bower_components/angular-qr/angular-qr.min.js"></script>
```

Add the **ja.qr** module as a dependency to your application module:

```js
var myAppModule = angular.module('MyApp', ['ja.qr']);
```

## Directive    

```html
<qr text="string"></qr>
```

```html
<qr type-number="8" correction-level="'M'" size="200" input-mode="'ALPHA_NUM'" text="string"></qr>
```

### Required attributes

#### text
Your text to encode

### Optional attributes

#### type-number
- 1-40
- default value: `0` = minimal required version

#### correction-level
- `L` - Low
- `M` - Medium (default)
- `Q` - Quartile
- `H` - High
 
#### size
Size in pixels
- default value: 250

#### input-mode
- `NUMBER`: *0, 1, 2, 3, 4, 5, 6, 7, 8, 9*
- `ALPHA_NUM`: *0–9, A–Z (upper-case only), space, $, %, *, +, -, ., /, :*
- `8bit`: *[ISO 8859-1](http://en.wikipedia.org/wiki/ISO_8859-1)*
- default value: minimal required input mode based on input text

# Contributing

Contributions are welcome. Please make a pull request against canary branch and do not bump versions. Also include tests.

# Testing

We use karma and jshint to ensure the quality of the code. The easiest way to run these checks is to use grunt:

    npm install -g grunt-cli
    npm install
    bower install
    grunt

The karma task will try to open Chrome as a browser in which to run the tests. Make sure this is available or change the configuration in `test/test.config.js` 


# License

The MIT License

Copyright (c) 2013 [Jan Antala](http://www.janantala.com)


CHANGELOG.md
<a name="v0.1.4"></a>
### v0.1.4 (2013-12-18)


#### Features

* **bower:** require angular 1.2.x version ([b5458a21](https://github.com/janantala/angular-qr/commit/b5458a2156cb00d1ea9e4520db17a07c1079b8ba))

<a name="v0.1.3"></a>
### v0.1.3 (2013-11-14)


#### Bug Fixes

* **is8bit:** max 8bit character has code 255 not 256 ([deebe0eb](https://github.com/janantala/angular-qr/commit/deebe0eb2eb353a52e023a37e1478e75f7b7221a))


<a name="v0.1.2"></a>
### v0.1.2 (2013-11-13)


#### Features

* **conventional-changelog:** add changelog support ([8dafd515](https://github.com/janantala/angular-qr/commit/8dafd515f0bf361366047838b024c22d767bae0b))



test/test.conf.js
module.exports = function(config) {
  config.set({
    basePath: '..',
    files: [
      'bower_components/angular/angular.js',
      'bower_components/angular-mocks/angular-mocks.js',
      'lib/qrcode.js',
      'src/angular-qr.js',
      'test/*.spec.js'
    ],
    frameworks: ['jasmine'],
    singleRun: true,
    browsers: [ 'Chrome' ]
  });
};

test/unit.spec.js
describe('ja.qr', function() {

  beforeEach(module('ja.qr'));

  // var elm, scope, ctrl;

  // beforeEach(inject(function($rootScope, $compile) {
  //   elm = angular.element('<qr type-number="0" correction-level="\'M\'" size="200" text="\'hello there\'"></qr>');
  //   scope = $rootScope;

  //   $compile(elm)(scope);
  //   scope.$digest();
  // }));
  
  describe('QrCtrl', function() {

    var scope, ctrl;
    beforeEach(inject(function($rootScope, $controller) {
      scope = $rootScope.$new();
      ctrl = $controller('QrCtrl', {$scope: scope });
    }));

    it('should detect NUMBER text', function() {
      expect(scope.isNUMBER('123')).toBe(true);
      expect(scope.isALPHA_NUM('123')).toBe(true);
      expect(scope.is8bit('123')).toBe(true);
    });

    it('should detect ALPHA_NUM text', function() {
      expect(scope.isNUMBER('A123')).toBe(false);
      expect(scope.isALPHA_NUM('A123')).toBe(true);
      expect(scope.is8bit('A123')).toBe(true);
    });

    it('should detect 8bit text', function() {
      expect(scope.isNUMBER('aA123')).toBe(false);
      expect(scope.isALPHA_NUM('aA123')).toBe(false);
      expect(scope.is8bit('aA123')).toBe(true);
    });

    it('should detect non 8bit text', function() {
      expect(scope.isNUMBER('čaA123')).toBe(false);
      expect(scope.isALPHA_NUM('čaA123')).toBe(false);
      expect(scope.is8bit('čaA123')).toBe(false);
    });

    it('should check text input mode', function() {
      expect(scope.checkInputMode('NUMBER', '123')).toBe(true);
      expect(scope.checkInputMode('ALPHA_NUM', '123')).toBe(true);
      expect(scope.checkInputMode('8bit', '123')).toBe(true);

      expect(scope.checkInputMode('ALPHA_NUM', 'A123')).toBe(true);
      expect(scope.checkInputMode('8bit', 'A123')).toBe(true);

      expect(scope.checkInputMode('8bit', 'aA123')).toBe(true);
    });

    it('should get the simpliest input mode', function() {
      expect(scope.getInputMode('123')).toBe('NUMBER');
      expect(scope.getInputMode('123')).not.toBe('ALPHA_NUM');
      expect(scope.getInputMode('123')).not.toBe('8bit');

      expect(scope.getInputMode('A123')).not.toBe('NUMBER');
      expect(scope.getInputMode('A123')).toBe('ALPHA_NUM');
      expect(scope.getInputMode('A123')).not.toBe('8bit');
      
      expect(scope.getInputMode('aA123')).not.toBe('NUMBER');
      expect(scope.getInputMode('aA123')).not.toBe('ALPHA_NUM');
      expect(scope.getInputMode('aA123')).toBe('8bit');

      expect(scope.getInputMode('A,123')).not.toBe('NUMBER');
      expect(scope.getInputMode('A,123')).not.toBe('ALPHA_NUM');
      expect(scope.getInputMode('A,123')).toBe('8bit');
    });

  });
});

Gruntfile.js
module.exports = function (grunt) {
  require('load-grunt-tasks')(grunt);

  // Default task.
  grunt.registerTask('default', ['karma', 'jshint']);
  grunt.registerTask('build', ['karma', 'jshint', 'concat', 'uglify']);

  var karmaConfig = function(configFile, customOptions) {
    var options = { configFile: configFile, keepalive: true };
    var travisOptions = process.env.TRAVIS && { browsers: ['Firefox'], reporters: 'dots' };
    return grunt.util._.extend(options, customOptions, travisOptions);
  };

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('bower.json'),
    meta: {
      banner: '/*!\n' +
        ' * <%= pkg.name %> v<%= pkg.version %>\n' +
        ' * The MIT License\n' +
        ' * Copyright (c) 2013 Jan Antala\n' +
        ' */'
    },
    uglify: {
      options: {
        preserveComments: 'some'
      },
      dist: {
        src: '<%= pkg.name %>.js',
        dest: '<%= pkg.name %>.min.js'
      }
    },
    concat: {
      options: {
        process: true,
        banner: '<%= meta.banner %>\n\n'
      },
      dist: {
        src: 'src/<%= pkg.name %>.js',
        dest: '<%= pkg.name %>.js'
      }
    },
    karma: {
      unit: {
        options: karmaConfig('test/test.conf.js')
      }
    },
    jshint:{
      files:['src/**/*.js', 'test/**/*.js'],
      options: {
        curly:true,
        eqeqeq:true,
        immed:true,
        latedef:true,
        newcap:true,
        noarg:true,
        sub:true,
        boss:true,
        eqnull:true,
        devel:true,
        globals:{}
      }
    },
    changelog: {
      options: {
        dest: 'CHANGELOG.md'
      }
    },
  });
};

angular-qr.js
/*!
 * angular-qr v0.1.4
 * The MIT License
 * Copyright (c) 2013 Jan Antala
 */

(function(QRCode){
  'use strict';

  angular.module('ja.qr', [])
  .controller('QrCtrl', ['$scope', function($scope){
    $scope.getTypeNumeber = function(){
      return $scope.typeNumber || 0;
    };

    $scope.getCorrection = function(){
      var levels = {
        'L': 1,
        'M': 0,
        'Q': 3,
        'H': 2
      };

      var correctionLevel = $scope.correctionLevel || 0;
      return levels[correctionLevel] || 0;
    };

    $scope.getText = function(){
      return $scope.text || '';
    };

    $scope.getSize = function(){
      return $scope.size || 250;
    };

    $scope.isNUMBER = function(text){
      var ALLOWEDCHARS = /^[0-9]*$/;
      return ALLOWEDCHARS.test(text);
    };

    $scope.isALPHA_NUM = function(text){
      var ALLOWEDCHARS = /^[0-9A-Z $%*+\-./:]*$/;
      return ALLOWEDCHARS.test(text);
    };

    $scope.is8bit = function(text){
      for (var i = 0; i < text.length; i++) {
        var code = text.charCodeAt(i);
        if (code > 255) {
          return false;
        }
      }
      return true;
    };

    $scope.checkInputMode = function(inputMode, text){
      if (inputMode === 'NUMBER' && !$scope.isNUMBER(text)) {
        throw new Error('The `NUMBER` input mode is invalid for text.');
      }
      else if (inputMode === 'ALPHA_NUM' && !$scope.isALPHA_NUM(text)) {
        throw new Error('The `ALPHA_NUM` input mode is invalid for text.');
      }
      else if (inputMode === '8bit' && !$scope.is8bit(text)) {
        throw new Error('The `8bit` input mode is invalid for text.');
      }
      else if (!$scope.is8bit(text)) {
        throw new Error('Input mode is invalid for text.');
      }

      return true;
    };

    $scope.getInputMode = function(text){
      var inputMode = $scope.inputMode;
      inputMode = inputMode || ($scope.isNUMBER(text) ? 'NUMBER' : undefined);
      inputMode = inputMode || ($scope.isALPHA_NUM(text) ? 'ALPHA_NUM' : undefined);
      inputMode = inputMode || ($scope.is8bit(text) ? '8bit' : '');

      return $scope.checkInputMode(inputMode, text) ? inputMode : '';
    };
  }])
  .directive('qr', ['$timeout', '$window', function($timeout, $window){

    return {
      restrict: 'E',
      template: '<canvas></canvas>',
      scope: {
        typeNumber: '=',
        correctionLevel: '=',
        inputMode: '=',
        size: '=',
        text: '='
      },
      controller: 'QrCtrl',
      link: function postlink(scope, element, attrs){

        if (scope.text === undefined) {
          throw new Error('The `text` attribute is required.');
        }

        var canvas = element.find('canvas')[0];
        var canvas2D = !!$window.CanvasRenderingContext2D;

        scope.TYPE_NUMBER = scope.getTypeNumeber();
        scope.TEXT = scope.getText();
        scope.CORRECTION = scope.getCorrection();
        scope.SIZE = scope.getSize();
        scope.INPUT_MODE = scope.getInputMode(scope.TEXT);

        var draw = function(context, qr, modules, tile){
          for (var row = 0; row < modules; row++) {
            for (var col = 0; col < modules; col++) {
              var w = (Math.ceil((col + 1) * tile) - Math.floor(col * tile)),
                  h = (Math.ceil((row + 1) * tile) - Math.floor(row * tile));
              context.fillStyle = qr.isDark(row, col) ? '#000' : '#fff';
              context.fillRect(Math.round(col * tile), Math.round(row * tile), w, h);
            }
          }
        };

        var render = function(canvas, value, typeNumber, correction, size, inputMode){
          var trim = /^\s+|\s+$/g;
          var text = value.replace(trim, '');

          var qr = new QRCode(typeNumber, correction, inputMode);
          qr.addData(text);
          qr.make();

          var context = canvas.getContext('2d');

          var modules = qr.getModuleCount();
          var tile = size / modules;
          canvas.width = canvas.height = size;

          if (canvas2D) {
            draw(context, qr, modules, tile);
          }
        };

        render(canvas, scope.TEXT, scope.TYPE_NUMBER, scope.CORRECTION, scope.SIZE, scope.INPUT_MODE);

        $timeout(function(){
          scope.$watch('text', function(value, old){
            if (value !== old) {
              scope.TEXT = scope.getText();
              scope.INPUT_MODE = scope.getInputMode(scope.TEXT);
              render(canvas, scope.TEXT, scope.TYPE_NUMBER, scope.CORRECTION, scope.SIZE, scope.INPUT_MODE);
            }
          });

          scope.$watch('correctionLevel', function(value, old){
            if (value !== old) {
              scope.CORRECTION = scope.getCorrection();
              render(canvas, scope.TEXT, scope.TYPE_NUMBER, scope.CORRECTION, scope.SIZE, scope.INPUT_MODE);
            }
          });

          scope.$watch('typeNumber', function(value, old){
            if (value !== old) {
              scope.TYPE_NUMBER = scope.getTypeNumeber();
              render(canvas, scope.TEXT, scope.TYPE_NUMBER, scope.CORRECTION, scope.SIZE, scope.INPUT_MODE);
            }
          });

          scope.$watch('size', function(value, old){
            if (value !== old) {
              scope.SIZE = scope.getSize();
              render(canvas, scope.TEXT, scope.TYPE_NUMBER, scope.CORRECTION, scope.SIZE, scope.INPUT_MODE);
            }
          });

          scope.$watch('inputMode', function(value, old){
            if (value !== old) {
              scope.INPUT_MODE = scope.getInputMode(scope.TEXT);
              render(canvas, scope.TEXT, scope.TYPE_NUMBER, scope.CORRECTION, scope.SIZE, scope.INPUT_MODE);
            }
          });
        });

      }
    };
  }]);

})(window.QRCode);
